### ISSUE TITLE
PHP SoapClient usage

### ISSUE THREAD

**Issue by jurchiks** [Travelport/travelport-uapi-tutorial-php#2]
Could you please provide a working code example using PHP SoapClient?

I've tried to make the Classic API work with SoapClient, but it fails with error `SoapFault exception: [HTTP] Could not connect to host` when trying to call this line: https://github.com/jurchiks/travelport-api/blob/master/test.php#L35

It seems the error does not change whether or not I use my company's provided authentication credentials, so it makes me think the problem is with something else, but the error message is far too short and does not give me any useful info at all.

I am planning to implement the universal API too, but I am not sure if I won't run into the same problem there as well, which is why I'd like to see a working example.

**Comment by vivekjyotipramanik** [comment]
Hi jurchiks,

Thanks you for your interest in Travelport Universal API. Could you please confirm the Credentials which you have is for Classic API or For Universal API. If it is for Universal API it might not work for Classic API.

The Universal API PHP sample working code can be found in this link, https://github.com/Travelport/travelport-uapi-tutorial-php.

Currently we do not have any sample ready for PHP for Classic API. We will work on the sample and provide as soon as possible in this thread. Thanks.

Thanks & Regards
Vivekjyoti Pramanik

**Comment by jurchiks** [comment]
All your example code uses curl, not SoapClient. That's the main problem, your APIs are SOAP APIs, yet you do not use a dedicated SOAP client to access it in your examples.

**Comment by vivekjyotipramanik** [comment]
Hi jurchiks,

We will provide an sample ith SoapClient as soon as possible. Thanks.

Thanks & Regards
Vivekjyoti Pramanik

**Comment by jurchiks** [comment]
Alright, thank you!

**Comment by jurchiks** [comment]
Any news? I really hope that PHP SoapClient actually works with your API, because otherwise I have to manually write a whole proprietary SOAP XML processing engine on my own, and I would hate to do that.

**Comment by jurchiks** [comment]
By the way, you have a typo in your Galileo Web Services documentation (for example, here: https://developer.travelport.com/euf/assets/developer-network/pdfs/Classic-API-brochure.pdf, but I've seen the same thing in other places as well).
It says that the Classic API uses SSLv3, but when using `'ssl_method' => SOAP_SSL_METHOD_SSLv3` in SoapClient $options, it does not work and I get a SoapFault with the message `[HTTP] Could not connect to host`. If I change it to `SOAP_SSL_METHOD_TLS`, then it shows me a proper error message, e.g. `[76] Authentication Result Error Message Response  (76): Authentication credentials are invalid.`.

**Comment by vivekjyotipramanik** [comment]
Hi jurchiks,

Thanks for the correction in Classic API documentation. I will provide the details to the concerned team to work with.

Regarding the PHP SoapClient, we are working on it and will provide an update soon. Thanks.

Thanks & Regards
Vivekjyoti Pramanik

**Comment by jurchiks** [comment]
I've gotten the Classic API to work with SoapClient.
If you download my project: https://github.com/jurchiks/travelport-api,
rename config-example.php to config.php, put your credentials in, and execute `php example_classic.php`, it will complete successfully, and I'm not using any dirty hacks to make it work.

However, I cannot get the Universal API to work because you require the username:password to be prefixed with "UniversalAPI/", which the PHP SoapClient does not support directly (and as far as I can see, something like that isn't mentioned in the SOAP standard either, and even your own code examples don't use the prefix: https://github.com/Travelport/travelport-uapi-tutorial-php/blob/master/Universal/phpSample_Universal.php#L45, https://github.com/Travelport/travelport-uapi-tutorial-c-sharp/blob/master/ConsoleApplication1/ConsoleApplication1/Utility/Helper.cs#L52, https://github.com/Travelport/travelport-uapi-tutorial-c-sharp/blob/master/ConsoleApplication1/ConsoleApplication1/Utility/HttpHeaderMessageInspector.cs#L52).
I tried just prefixing the username with that, but the code as seen in my repository fails with or without the prefix.

**Comment by vivekjyotipramanik** [comment]
Hi jurchiks,

The prefix to the Username is required and for other code base and current PHP code base it needs to be use as part of the Username itself. The Universal API/ prefix is required to pass the security in uAPI. In the current Github code base we say, 'Please use the UserName which you received in the welcome letter:Please use the password which you received in the welcome letter' and we expect developers to use the Username with "Universal API/" as prefix.
Could you please confirm what issue you are getting while using 'Universal API/'. Thanks.

Thanks & Regards
Vivekjyoti Pramanik

**Comment by jurchiks** [comment]
If I execute `php example_universal.php` in my project, it throws an exception with the message `[WSDL] SOAP-ERROR: Parsing WSDL: Couldn't load from 'https://americas.universal-api.pp.travelport.com/B2BGateway/connect/uAPI/TerminalService' : failed to load external entity "https://americas.universal-api.pp.travelport.com/B2BGateway/connect/uAPI/TerminalService"
`.
The exact same error is shown if I change `$config['username']` to `'UniversalAPI/' . $config['username']`.

The credentials are definitely correct, they work with the Classic API, and the URL seems to be correct as well (according to https://support.travelport.com/webhelp/uapi/Content/Getting_Started/Endpoints_Services.htm), but the error message doesn't give me any useful info, as usual with SOAP.

**Comment by vivekjyotipramanik** [comment]
Hi jurchiks,

Could you please try your credentials in the link https://demo.travelportuniversalapi.com/Home.
In the top right corner,  look for 'CREDENTIALS' and click on Production, and you can change it to custom Credentials. Please put your Credentials in the input area and try it out, if it is working there. The Classic API Credentials and uAPI Credentials might be different. You can try with simple Air search. Please check in the link if it is working fine there. please let me know how it goes. Thanks.

Thanks & Regards
Vivekjyoti Pramanik

**Comment by jurchiks** [comment]
It says `Invalid Credentials`. Not sure what to put in the Branch field though, but it's not marked as required...

**Comment by vivekjyotipramanik** [comment]
Hi jurchiks,

As I have said, Classic API and Universal API is completely different and the so are the credentials. In Universal API credentials, BranchCode/TragetBranch is a Critical part. Please raise a request for Universal API Credentials from the below link. Please let me know how it goes. Thanks.

https://developer.travelport.com/app/developer-network/developer-network

Search for: 'Request Test Access for Universal API'

Thanks & Regards
Vivekjyoti Pramanik

**Comment by jurchiks** [comment]
Unfortunately the Classic API credentials do not work in this page: https://developer.travelport.com/app/utils/login_form_dev
so it seems that I can't do that. I don't have those credentials, and as far as I know, only someone much higher up than me has them; they're certainly not in any files in the code base.
Looks like this could take a while.

**Comment by vivekjyotipramanik** [comment]
Hi jurchiks,

You can register in this page and ask for Universal API Test access. Please find below the register link. Please let me know how it goes. Thanks.

https://developer.travelport.com/app/utils/login_form_dev

Please use CREATE A NEW ACCOUNT


Thanks & Regards
Vivekjyoti Pramanik

**Comment by jurchiks** [comment]
Well, I would like to use an existing company account instead of creating my own; the company will need one for production use anyway.

**Comment by jurchiks** [comment]
By the way, I have to ask: in this page - https://support.travelport.com/webhelp/uapi/Content/Getting_Started/Easy_Overview/Complete_Services_List.htm - you have several services with identical names, but different WSDLs (as an example, let's take AirService).
According to this page: https://support.travelport.com/webhelp/uapi/Content/Getting_Started/Endpoints_Services.htm
the service name should to at the end of the end-point URL, e.g. 	
https://americas.universal-api.pp.travelport.com/B2BGateway/connect/uAPI/AirService
However, since there are two services with the same name (Air/AirService and Universal/AirService), how am I supposed to make the distinction between the two? In the Classic API, the service name is the WSDL name, but here...

**Comment by vivekjyotipramanik** [comment]
Hi jurchiks,

The distinction can be made in the request XML based on the TargetNameSpace. If you go through the different xml request/response samples, if the AirService is in Air, the TargetNameSpace of the request will be Air, like air:AirPriceReq and air:AirMerchandisingOfferAvailabilityReq etc. If the AirService is in Universal we need to use TargetnameSpace with universal, like univ:AirCreatereservationreq etc. Please let me know if that clarify the question. Thanks.

https://support.travelport.com/webhelp/uapi/uAPI.htm#Air/Air_Booking/Create_Air_Booking/Creating_Air_Bookings.htm%3FTocPath%3DAir%7CAir%2520Shopping%2520and%2520Booking%7CAir%2520Booking%7CCreating%2520Air%2520Bookings%7C_____0

Thanks & Regards
Vivekjyoti Pramanik

**Comment by jurchiks** [comment]
It seems that your Universal API does not work with WSDL the way Galileo does.
The service URL is supposed to return a WSDL with all the available requests and their responses.
The Universal API duplicate name issue must mean that this is impossible, because the same URL is used for two different services, each with their own WSDL.

I'm guessing that your Universal API is not actually in WSDL mode, i.e. I have to use it like so:
$client = new SoapClient(null, array(
    'location' => 'https://americas.universal-api.pp.travelport.com/B2BGateway/connect/uAPI/AirService',
    'uri' => 'http://www.travelport.com/service/air_v35_0', // targetNamespace in Air.wsdl
));
Or, alternatively, I have to download your WSDL files and use them locally:
$client = new SoapClient(__DIR__ . '/path/to/Air.wsdl');

Neither of these is maintainable in long-term as I have to either manually update all the WSDLs on every change (thus risking breaking any live site if I forget to do so in time) or update the version numbers in code.

**Comment by jurchiks** [comment]
Another documentation bug:
This page lists Reporting/UniversalRecordReportService in Reporting.wsdl
https://support.travelport.com/webhelp/uapi/Content/Getting_Started/Easy_Overview/Complete_Services_List.htm
But there is no such WSDL file in this archive: https://support.travelport.com/webhelp/uapi/Content/Resources/uAPI_WSDLschema_Release-V15.5.0.91.zip
and the service is also not mentioned here:
https://support.travelport.com/webhelp/uapi/uAPI.htm#Schemas/Schema_Overview.htm%3FTocPath%3DSchema%7C_____0

**Comment by vivekjyotipramanik** [comment]
Hi jurchiks,

We are working on to resolve the documentation issue. Thanks.

Thanks & Regards
Vivekjyoti Pramanik

**Comment by pancid** [comment]
Hi jurchiks and vivekjyotipramanik,

I have also been trying to get the PHP internal SOAP client to work with uAPI, but without any initial success. 

Have you been able to implent a uAPI SOAP client by now?

**Comment by jurchiks** [comment]
This is the latest version of my API that I could make: http://www.4shared.com/zip/Pc_BUI8pba/travelport-api.html
I was forced to remove it from github by my boss, who apparently doesn't like open-sourcing stuff like this, even though it is in no way connected to the work I was doing.
At least this version can actually dump out the WSDL function list, but I don't think I ever finished the calls to those functions.

**Comment by pancid** [comment]
Thanks Juris. Very much appreciated. I am still parsing the PHP results based on the official Travelport example here on GitHub, but may look into that again at a later stage.

**Comment by jurchiks** [comment]
Yeah, the official travelport example doesn't use SoapClient and that is really bad. This is a SOAP service, so it should use some kind of OOP for all the error checking and data conversion. Even their own proprietary library would be better than this.

**Comment by hshhhhh** [comment]
> Yeah, the official travelport example doesn't use SoapClient and that is really bad. This is a SOAP service, so it should use some kind of OOP for all the error checking and data conversion. Even their own proprietary library would be better than this.

It looks like problem in php. I tried to make ping with soap and it looks like in system.wsdl their have few methods with the same name 'service' to call from soapclient and it should call differents methods based on classname of request we send into it, but in real life PHP just called first method, so when you trying to do:

```php
class PingReq
{
	public $Payload = null;
	public function __construct($Payload)
	{
		$this->Payload = $Payload;
	}
}

$query = new PingReq('test');
$response = $client->service($query);
```
it called not SystemPingPortType but ExternalCacheAccessBinding

Here is php bug for it (7 fricking years!!!1): https://bugs.php.net/bug.php?id=49169

**Comment by jurchiks** [comment]
Well, the bug offers a workaround: http://us3.php.net/manual/en/soapserver.handle.php#81750

**Comment by hshhhhh** [comment]
@jurchiks but I don't think that it's a good idea to rewrite all wsdl's to rpc/literal style and also I'm not sure that it would helps because it looks like php-core-team do not plan to support 'method overriding' for soap.

So unfortunatelly I will use soap via curl, god bless php!

**Comment by jurchiks** [comment]
Yeah, of course it's not a good idea, especially in the case of proprietary WSDLs. It is ok for personal WSDLs tho.

**Comment by hshhhhh** [comment]
@vivekjyotipramanik any solution?

**Comment by jurchiks** [comment]
They don't even officially support PHP SOAP, they'll just tell you to use curl.

**Comment by hshhhhh** [comment]
@jurchiks  thank you a lot for you help. I will try to play with your examples.

**Comment by jurchiks** [comment]
Good luck!

**Comment by hshhhhh** [comment]
@jurchiks  if I understand correctly you are sending xml requests via soap, but how do you generate that xml? by hands? or do you use any xsd2xml or wsd2xml? 

Do I correct?

**Comment by jurchiks** [comment]
As you can see in the code I uploaded to 4shared, I did not get the Universal API to work with SOAP.
For Classic API, however, it is as simple as constructing a normal array with the same structure as the XML (see src\classic\services\XmlSelect.php). For more complex structures, you just make nested arrays, and it should work (I didn't test any complex structures though).

**Comment by phpdevelopers** [comment]
any update?
Can anybody soapclient use?

**Comment by hshhhhh** [comment]
No, it not possible because PHP SoapClient doesn't support 'method overloading' what using in that travelport java soap server. So write client for php with method overloading support or generate raw xml and use cURL  :(

**Comment by semsphy** [comment]
+1

**Comment by boryashkin** [comment]
**It is possible**, btw, but requires some changes in wsdl's becauseof unsupported 'method overloading'.
